src/engine/ContextDependentPiecewiseLinearConstraint.h

```diff
@@ -0,0 +1,246 @@
+/*********************                                                        */
+/*! \file ContextDependentPiecewiseLinearConstraint.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aleksandar Zeljic
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** This class is an implementation of PiecewiseLinearConstraint using
+ ** context-dependent (CD) data structures from CVC4, which lazily and automatically
+ ** backtrack in sync with the central context object. These data structures
+ ** are delicate and require that data stored is safe to memcopy around.
+ **
+ **
+ ** Members are initialized using the initializeCDOs( Context context ) method.
+ ** Subclasses need to call initializeDuplicatesCDOs() method in the
+ ** duplicateConstraint() method to avoid sharing context-dependent members.
+ ** CDOs need not be initialized for pre-processing purposes.
+ **
+ ** The class uses CDOs to record:
+ **  * _cdConstraintActive
+ **  * _cdPhaseStatus
+ **  * _cdInfeasibleCases
+ **
+ ** PhaseStatus enumeration is used to keep track of current phase of the
+ ** object, as well as to record previously explored phases. This replaces
+ ** moving PiecewiseLinearCaseSplits around.
+ **
+ ** ContextDependentPiecewiseLinearConstraint stores locally which phases have
+ ** been explored so far using _cdInfeasibleCases. To communicate with the
+ ** search it implements two methods:
+ **   * markInfeasible( PhaseStatus caseId ) - marks explored phase
+ **   * nextFeasibleCase() - obtains next unexplored case if one exists
+ **
+ ** These methods rely on subclass properly handling:
+ **  * _numCases field, which denotes the number of piecewise linear
+ **     segments the constraint has, initialized by the constructor.
+ **  * getAllCases() virtual method returns all possible phases represented using
+ **    PhaseStatus enumeration. The order of returned cases will determine the
+ **    search order, so this method should implement any search heuristics.
+ **/
+
+#ifndef __ContextDependentPiecewiseLinearConstraint_h__
+#define __ContextDependentPiecewiseLinearConstraint_h__
+
+#include "FloatUtils.h"
+#include "ITableau.h"
+#include "List.h"
+#include "Map.h"
+#include "MarabouError.h"
+#include "PiecewiseLinearCaseSplit.h"
+#include "PiecewiseLinearConstraint.h"
+#include "PiecewiseLinearFunctionType.h"
+#include "Queue.h"
+#include "Tightening.h"
+#include "context/cdlist.h"
+#include "context/cdo.h"
+#include "context/context.h"
+
+class Equation;
+class BoundManager;
+class ITableau;
+class InputQuery;
+class String;
+
+class ContextDependentPiecewiseLinearConstraint
+   : public virtual PiecewiseLinearConstraint
```

Comment:
is virtual inheritance needed here?

Reply:
No, especially since the goal is to merge the two classes eventually. Fixed.